driver/src/main/com/mongodb/client/gridfs/GridFSBucketImpl.java

```diff
@@ -149,26 +149,16 @@ public ObjectId uploadFromStream(final String filename, final InputStream source
         int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
         byte[] buffer = new byte[chunkSize];
         int len;
-        MongoGridFSException savedThrowable = null;
         try {
             while ((len = source.read(buffer)) != -1) {
                 uploadStream.write(buffer, 0, len);
             }
+            uploadStream.close();
         } catch (IOException e) {
-            savedThrowable = new MongoGridFSException("IO Exception when reading from the InputStream", e);
+            uploadStream.abort();
+            throw new MongoGridFSException("IOException when reading from the InputStream", e);
         } catch (Exception e) {
-            savedThrowable = new MongoGridFSException("Unexpected exception when reading stream and writing to GridFS", e);
-        } finally {
-            try {
-                uploadStream.close();
-            } catch (Exception e) {
-                if (savedThrowable == null) {
-                    throw new MongoGridFSException("Unexpected exception when closing the GridFSUploadStream", e);
-                }
-            }
-            if (savedThrowable != null) {
-                throw savedThrowable;
-            }
+            throw new MongoGridFSException("Unexpected exception when reading stream and writing to GridFS", e);
```

Comment:
Because according to the Spec: 

> Abort MUST raise an error if it is unable to succesfully abort the upload (for example, if an error occurs while deleting any chunks already uploaded). However, if the upload is being aborted because the source stream provided to upload_from_stream raised an error then the original error should be re-raised.

So I read that as, the only exception that we should abort on is the `IOException` due to the source stream failing and any other exception should be re raised.